File: 0087-scramble-string/0087-scramble-string.java
class Solution {
    Map<String,Boolean> a=new HashMap<String,Boolean>();
    
    public boolean isScramble(String s1, String s2) {
        if (s1.length()!=s2.length()) return false;
        if (s1.equals(s2)) return true;
        return fun(s1,s2);
    }
    
    boolean fun(String s1,String s2){
        int n=s1.length();
        String k=s1+"#"+s2;
        
        //System.out.println(s1+" "+s2);
        
        if (s1.equals(s2)){
            a.put(k,true);
            return true;
        }
        else if (a.containsKey(k))  return a.get(k);
        else if(n==1) return false;
        
        for(int i=1;i<=n-1;i++){
            //System.out.println(b1+" "+b2+" "+b3+" "+b4);
            
            if((fun(s1.substring(0,i),s2.substring(0,i))) && (fun(s1.substring(i),s2.substring(i)))){
                a.put(k,true);
                return true;
            }
            if((fun(s1.substring(0,i),s2.substring(n-i))) && (fun(s1.substring(i),s2.substring(0,n-i)))){
                a.put(k,true);
                return true;
            }
            
        }
        a.put(k,false);
        return false;
    }
    
}

File: 0087-scramble-string/0087-scramble-string.py
class Solution:
    
    def isScramble(self, s1: str, s2: str) -> bool:
        mm={}
        if len(s1)!=len(s2):
            return False
        if s1==s2:
            return True
        if len(s1)==1:
            return s1==s2
        
        def tra(s1, s2):
            n=len(s1)
            print(s1, s2, sep=' ')
            if s1==s2:
                mm[s1,s2]=True
                return True
            elif (s1,s2) in mm:
                return mm[s1,s2]
            elif n==1:
                return False
            for i in range (1,len(s1)):
                if (tra(s1[:i],s2[:i])  and  tra(s1[i:],s2[i:])) :
                    mm[s1,s2]=True
                    return True
            
                if (tra(s1[0:i],s2[-i:]) and  tra(s1[i:],s2[0:n-i])) :
                    mm[s1,s2]=True
                    return True
                
            mm[s1,s2]=False
            return mm[s1,s2]
        
        return tra(s1,s2)
        
    
    
                
    
        